Pyk library: Library for handling frontend JSON KAST terms programatically

ehildenb opened the pull request:
This introduces the pyk library, which is being used by (1) the evm-coverage repo, and (2) the beacon-chain repo. Currently there are PRs into those repos waiting on this PR.

I have added a simple test-harness as well, and tested that the pyk library can generate well-formatted K pretty format terms that `kast` can parse. We can add more tests as features/bugs are uncovered I figure.

ehildenb committed:
.gitignore, lib/py-kframework: initial python library for KAST manip

ehildenb committed:
pyk: re-shuffle files

ehildenb committed:
lib/py-kframework: add KAST forms for KRequire and KDefinition, simplify spec writing/proving methods

ehildenb committed:
k-distribution/pyk/kast.py: print one-line cells on single line

ehildenb committed:
k-distribution/kast.py: add empty Map klabel to symbol table

ehildenb committed:
k-distribution/tests/pyk: add simple tests of pyk library

ehildenb committed:
k-distribution/kast.py: add assocWithUnit printer, better logic for multi-line cells

ehildenb committed:
k-distribution/tests/pyk: passing pyk tests

ehildenb committed:
Dockerfile.debian: install python3

dwightguth commented on k-distribution/src/main/scripts/lib/py-kframework/kast.py:
```diff
@@ -0,0 +1,228 @@
+                   , klabelEmptyK  : (lambda : ".")
+                   }
+
+BOOL_expressions = { "_andBool__BOOL" : binOpStr("\nandBool")
```
this is not robust. relying on the klabels generated for productions that don't have the klabel and symbol attributes is going to be fragile and in particular it will break in the llvm and haskell backends. if you want to refer to all of these labels explicitly, let's give them explicit klabels and the symbol attribute.

ehildenb replied:
I've added explicit `klabel` and `symbol` attributes for many parts of `domains.k`, at least all the parts that `pyk` understands for now.

dwightguth commented on k-distribution/src/main/scripts/lib/py-kframework/pyk.py:
```diff
@@ -0,0 +1,240 @@
+
+from kast import *
+
+def match(pattern, kast):
```
what is this being used for? it doesn't seem to deal with sorts very intelligently.

ehildenb replied:
No, it certainly doesn't. It's not meant to be a robust pattern matcher.

It's used for doing simple rewrites/manip over the AST before handing it over to K. Often it's easier to specify the desired behavior as a rewrite (which uses the matcher) than to manually traverse/update the term.

This is used in function `normalizeKLabels`, for example, which does things like replacing `true and TERM` with just `TERM`.

ehildenb committed:
k-distribution/pyk.py: more sensible function definition order

ehildenb committed:
k-distribution/{domains.k,kast.py}: explicit klabels for builtins recognized by kast.py

ehildenb committed:
regression-new/fatalWarnings: update output

dwightguth reviewed (changes requested):
I still see several references to outdated _BOOL klabels.

ehildenb committed:
kore/outer.scala: care about `symbol` attributes for Production.equals

ehildenb committed:
k-distribution/{kast.py,pyk.py}: correct klabels to use

ehildenb committed:
k-distribution/tests/pyk: more thorough testing harness

ehildenb commented:
I've added a more useful testing harness, which catches this issue with `normalizeKLabels` having the incorrect klabels from the `domains.k`.

I've also added `symbol` to the attributes which `Production` uses when testing `equals`.

ehildenb committed:
k-distribution/pyk: correct more klabels, simplify out trivial requires clauses

ehildenb committed:
k-distribution/tests/pyk: add test of proving functionality

ehildenb committed:
Merge branch 'master' into pyk-library
